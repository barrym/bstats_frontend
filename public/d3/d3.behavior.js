(function(){d3.behavior = {};
// TODO unbind zoom behavior?
// TODO unbind listener?
d3.behavior.zoom = function() {
  var x = 0,
      y = 0,
      z = 0,
      panning,
      zooming,
      locations = {}, // identifier -> location
      last = 0,
      target,
      targuments,
      event = d3.dispatch("zoom");

  // mousewheel events are totally broken!
  // https://bugs.webkit.org/show_bug.cgi?id=40441
  // not only that, but Chrome and Safari differ in re. to acceleration!
  var div = d3.select("body").append("div")
      .style("visibility", "hidden")
      .style("top", 0)
      .style("height", 0)
      .style("width", 0)
      .style("overflow-y", "scroll")
    .append("div")
      .style("height", "2000px")
    .node().parentNode;

  function zoom() {
    this
        .on("mousedown.zoom", mousedown)
        .on("mousewheel.zoom", mousewheel)
        .on("DOMMouseScroll.zoom", dblclick)
        .on("dblclick.zoom", dblclick)
        .on("touchstart.zoom", touchstart);

    d3.select(window)
        .on("mousemove.zoom", mousemove)
        .on("mouseup.zoom", mouseup)
        .on("touchmove.zoom", touchmove)
        .on("touchend.zoom", touchup);
  }

  function location(point) {
    return [point[0] - x, point[1] - y, z];
  }

  // snapshot the target, data and index for subsequent dispatch
  function start() {
    target = this;
    targuments = arguments;
  }

  // Note: Since we don't rotate, it's possible for the touches to become
  // slightly detached from their original positions. Thus, we recompute the
  // touch points on touchend as well as touchstart!
  function touchup() {
    var touches = d3.svg.touches(target),
        i = -1,
        n = touches.length,
        touch;
    while (++i < n) locations[(touch = touches[i]).identifier] = location(touch);
    return touches;
  }

  function touchstart() {
    start.apply(this, arguments);
    var touches = touchup(),
        touch,
        now = Date.now();

    // doubletap detection
    if ((touches.length === 1) && (now - last < 300)) {
      var touch = touches[0];
      zoomto(1 + Math.floor(z), touch, locations[touch.identifier]);
    }
    last = now;
  }

  function touchmove() {
    var touches = d3.svg.touches(target);
    switch (touches.length) {

      // single-touch pan
      case 1: {
        var touch = touches[0];
        zoomto(z, touch, locations[touch.identifier]);
        break;
      }

      // double-touch pan + zoom
      case 2: {
        var p0 = touches[0],
            p1 = touches[1],
            p2 = [(p0[0] + p1[0]) / 2, (p0[1] + p1[1]) / 2],
            l0 = locations[p0.identifier],
            l1 = locations[p1.identifier],
            l2 = [(l0[0] + l1[0]) / 2, (l0[1] + l1[1]) / 2, l0[2]];
        zoomto(Math.log(d3.event.scale) / Math.LN2 + l0[2], p2, l2);
        break;
      }
    }
  }

  function mousedown() {
    start.apply(this, arguments);
    panning = location(d3.svg.mouse(target));
    d3.event.preventDefault();
    window.focus();
  }

  function mousemove() {
    zooming = null;
    if (panning) zoomto(z, d3.svg.mouse(target), panning);
  }

  function mouseup() {
    if (panning) {
      mousemove();
      panning = null;
    }
  }

  function mousewheel() {
    start.apply(this, arguments);

    // store starting mouse location
    if (!zooming) zooming = location(d3.svg.mouse(target));

    // detect the pixels that would be scrolled by this wheel event
    var e = d3.event, delta;
    try {
      div.scrollTop = 1000;
      div.dispatchEvent(e);
      delta = 1000 - div.scrollTop;
    } catch (error) {
      delta = e.wheelDelta || -e.detail;
    }

    zoomto(delta * .005 + z, d3.svg.mouse(target), zooming);
  }

  function dblclick() {
    start.apply(this, arguments);
    var mouse = d3.svg.mouse(target);
    zoomto(d3.event.shiftKey ? Math.ceil(z - 1) : Math.floor(z + 1), mouse, location(mouse));
  }

  function zoomto(z1, x0, x1) {
    var k = Math.pow(2, (z = z1) - x1[2]);
    x = x0[0] - k * x1[0];
    y = x0[1] - k * x1[1];
    dispatch();
  }

  function dispatch() {
    var o = d3.event, // Events can be reentrant (e.g., focus).
        k = Math.pow(2, z);

    d3.event = {
      scale: k,
      translate: [x, y],
      transform: function(sx, sy) {
        if (sx) transform(sx, x);
        if (sy) transform(sy, y);
      }
    };

    function transform(scale, o) {
      var domain = scale.__domain || (scale.__domain = scale.domain()),
          range = scale.range().map(function(v) { return (v - o) / k; });
      scale.domain(domain).domain(range.map(scale.invert));
    }

    try {
      event.zoom.dispatch.apply(target, targuments);
    } finally {
      d3.event = o;
    }

    o.preventDefault();
  }

  zoom.on = function(type, listener) {
    event[type].add(listener);
    return zoom;
  };

  return zoom;
};
})();
