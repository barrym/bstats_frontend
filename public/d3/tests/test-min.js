require("./../lib/env-js/envjs/node");
require("./../d3");

function min(array) {
  var min = d3.min(array);
  return typeof min === "string" ? "\"" + min + "\"" : min;
}

console.log("min:");
console.log("                                [1] -> " + min([1]));
console.log("                    [5, 1, 2, 3, 4] -> " + min([5, 1, 2, 3, 4]));
console.log("                    [\"a\", \"b\", \"c\"] -> " + min(["c", "a", "b"]));
console.log("                        [\"20\", \"3\"] -> " + min(["20", "3"]));
console.log("                        [\"3\", \"20\"] -> " + min(["3", "20"]));
console.log("                            [20, 3] -> " + min([20, 3]));
console.log("                            [3, 20] -> " + min([3, 20]));
console.log("");

console.log("min of empty array is undefined:");
console.log("                                 [] -> " + min([]));
console.log("                             [null] -> " + min([null]));
console.log("                        [undefined] -> " + min([undefined]));
console.log("                              [NaN] -> " + min([NaN]));
console.log("                         [NaN, NaN] -> " + min([NaN, NaN]));
console.log("");

console.log("min ignores null, undefined, and NaN:");
console.log("               [NaN, 1, 2, 3, 4, 5] -> " + min([NaN, 1, 2, 3, 4, 5]));
console.log("               [1, 2, 3, 4, 5, NaN] -> " + min([1, 2, 3, 4, 5, NaN]));
console.log("   [10, null, 3, undefined, 5, NaN] -> " + min([10, null, 3, undefined, 5, NaN]));
console.log("");

console.log("min compares heterogenous types as numbers:");
console.log("                          [20, \"3\"] -> " + min([20, "3"]));
console.log("                          [\"20\", 3] -> " + min(["20", 3]));
console.log("                          [3, \"20\"] -> " + min([3, "20"]));
console.log("                          [\"3\", 20] -> " + min(["3", 20]));
console.log("");

console.log("min applies accessor function:");
console.log("  [1, 2, 3, 4, 5], [2, 4, 6, 8, 10] -> " + d3.min([[1, 2, 3, 4, 5], [2, 4, 6, 8, 10]], function(d) { return d3.max(d); }));
console.log("                    [1, 2, 3, 4, 5] -> " + d3.min([1, 2, 3, 4, 5], function(d, i) { return i; }));
console.log("");
