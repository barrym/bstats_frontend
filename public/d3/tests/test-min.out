min:
                                [1] -> 1
                    [5, 1, 2, 3, 4] -> 1
                    ["a", "b", "c"] -> "a"
                        ["20", "3"] -> "20"
                        ["3", "20"] -> "20"
                            [20, 3] -> 3
                            [3, 20] -> 3

min of empty array is undefined:
                                 [] -> undefined
                             [null] -> undefined
                        [undefined] -> undefined
                              [NaN] -> undefined
                         [NaN, NaN] -> undefined

min ignores null, undefined, and NaN:
               [NaN, 1, 2, 3, 4, 5] -> 1
               [1, 2, 3, 4, 5, NaN] -> 1
   [10, null, 3, undefined, 5, NaN] -> 3

min compares heterogenous types as numbers:
                          [20, "3"] -> "3"
                          ["20", 3] -> 3
                          [3, "20"] -> 3
                          ["3", 20] -> "3"

min applies accessor function:
  [1, 2, 3, 4, 5], [2, 4, 6, 8, 10] -> 5
                    [1, 2, 3, 4, 5] -> 0

